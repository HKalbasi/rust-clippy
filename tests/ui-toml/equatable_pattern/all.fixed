// run-rustfix

#![allow(unused_variables, dead_code, clippy::redundant_pattern_matching, clippy::op_ref)]
#![warn(clippy::equatable_if_let, clippy::equatable_matches)]

#[derive(PartialEq)]
enum Enum {
    TupleVariant(i32, u64),
    RecordVariant { a: i64, b: u32 },
    UnitVariant,
    Recursive(Struct),
}

#[derive(PartialEq)]
struct Struct {
    a: i32,
    b: bool,
}

#[derive(Clone, Copy)]
enum NotPartialEq {
    A,
    B,
}

#[derive(Clone, Copy)]
enum NotStructuralEq {
    A,
    B,
}

impl PartialEq for NotStructuralEq {
    fn eq(&self, _: &NotStructuralEq) -> bool {
        false
    }
}

#[derive(PartialEq)]
enum Generic<A, B> {
    VA(A),
    VB(B),
    VC,
}

#[derive(PartialEq)]
struct Generic2<A, B> {
    a: A,
    b: B,
}

fn main() {
    let a = 2;
    let b = 3;
    let c = Some(2);
    let d = Struct { a: 2, b: false };
    let e = Enum::UnitVariant;
    let f = NotPartialEq::A;
    let g = NotStructuralEq::A;
    let h: Generic<Enum, NotPartialEq> = Generic::VC;
    let i: Generic<Enum, NotStructuralEq> = Generic::VC;
    let j = vec![1, 2, 3, 4];
    let k = Some(&false);
    let l = Generic2 {
        a: Generic2 { a: "xxxx", b: 3 },
        b: Generic2 {
            a: &Enum::UnitVariant,
            b: false,
        },
    };
    let m = Generic2 { a: 3, b: 5 };
    let n = Some("xxxx");
    let mut o = j.iter();

    // true

    if c == Some(2) {}
    if d == (Struct { a: 2, b: false }) {}
    if e == Enum::TupleVariant(32, 64) {}
    if e == (Enum::RecordVariant { a: 64, b: 32 }) {}
    if (e, &d) == (Enum::UnitVariant, &Struct { a: 2, b: false }) {}
    if i == Generic::VA(Enum::UnitVariant) {}
    if j[1..3] == [7, 5] {}
    if j[..] == [1, 2, 3, 4] {}
    if k == Some(&true) {}
    if k == Some(&false) {}
    if &k == &Some(&true) {}
    if &&k == &&Some(&false) {}
    if l == (Generic2 { a: Generic2 { a: "yyy", b: 3 }, b: Generic2 { a: &Enum::UnitVariant, b: false } })
    {}
    if m == (Generic2 { a: 3, b: 5 }) {}
    if n == Some("yyy") {}
    let _ = c == Some(2);

    while o.next() == Some(&2) {}

    // false

    if let 2 | 3 = a {}
    if let x @ 2 = a {}
    if let Some(3 | 4) = c {}
    if let Struct { a, b: false } = d {}
    if let Struct { a: 2, b: x } = d {}
    if let NotPartialEq::A = f {}
    if let NotStructuralEq::A = g {}
    if let Some(NotPartialEq::A) = Some(f) {}
    if let None = Some(f) {}
    if let Some(NotStructuralEq::A) = Some(g) {}
    if let Generic::VA(Enum::UnitVariant) = h {}
    if let Generic::VB(NotPartialEq::A) = h {}
    if let Generic::VC = h {}
    if let Generic::VB(NotStructuralEq::A) = i {}
    if let [7, _] = j[2..] {}
    if let [1, 2 | 5, 3, 4] = j[..] {}
    if let [2, ..] = j[..] {}

    let _ = matches!(c, Some(x));
    let _ = matches!(c, Some(x) if x == 2);
    let _ = matches!(c, Some(2) if 3 > 5);

    while let Some(4 | 7) = o.next() {}
}
