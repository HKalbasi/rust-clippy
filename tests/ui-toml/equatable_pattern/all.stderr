error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:76:8
   |
LL |     if let Some(2) = c {}
   |        ^^^^^^^^^^^^^^^ help: try: `c == Some(2)`
   |
   = note: `-D clippy::equatable-if-let` implied by `-D warnings`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:77:8
   |
LL |     if let Struct { a: 2, b: false } = d {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `d == (Struct { a: 2, b: false })`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:78:8
   |
LL |     if let Enum::TupleVariant(32, 64) = e {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `e == Enum::TupleVariant(32, 64)`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:79:8
   |
LL |     if let Enum::RecordVariant { a: 64, b: 32 } = e {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `e == (Enum::RecordVariant { a: 64, b: 32 })`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:80:8
   |
LL |     if let (Enum::UnitVariant, &Struct { a: 2, b: false }) = (e, &d) {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `(e, &d) == (Enum::UnitVariant, &Struct { a: 2, b: false })`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:81:8
   |
LL |     if let Generic::VA(Enum::UnitVariant) = i {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `i == Generic::VA(Enum::UnitVariant)`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:82:8
   |
LL |     if let [7, 5] = j[1..3] {}
   |        ^^^^^^^^^^^^^^^^^^^^ help: try: `j[1..3] == [7, 5]`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:83:8
   |
LL |     if let [1, 2, 3, 4] = j[..] {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `j[..] == [1, 2, 3, 4]`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:84:8
   |
LL |     if let Some(true) = k {}
   |        ^^^^^^^^^^^^^^^^^^ help: try: `k == Some(&true)`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:85:8
   |
LL |     if let Some(&false) = k {}
   |        ^^^^^^^^^^^^^^^^^^^^ help: try: `k == Some(&false)`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:86:8
   |
LL |     if let Some(true) = &k {}
   |        ^^^^^^^^^^^^^^^^^^^ help: try: `&k == &Some(&true)`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:87:8
   |
LL |     if let Some(false) = &&k {}
   |        ^^^^^^^^^^^^^^^^^^^^^ help: try: `&&k == &&Some(&false)`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:88:8
   |
LL |       if let Generic2 {
   |  ________^
LL | |         a: Generic2 { a: "yyy", b: 3 },
LL | |         b: Generic2 {
LL | |             a: Enum::UnitVariant,
LL | |             b: false,
LL | |         },
LL | |     } = l
   | |_________^ help: try: `l == (Generic2 { a: Generic2 { a: "yyy", b: 3 }, b: Generic2 { a: &Enum::UnitVariant, b: false } })`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:96:8
   |
LL |     if let Generic2 { a: 3, b: 5 } = m {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `m == (Generic2 { a: 3, b: 5 })`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:97:8
   |
LL |     if let Some("yyy") = n {}
   |        ^^^^^^^^^^^^^^^^^^^ help: try: `n == Some("yyy")`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:98:13
   |
LL |     let _ = matches!(c, Some(2));
   |             ^^^^^^^^^^^^^^^^^^^^ help: try: `c == Some(2)`
   |
   = note: `-D clippy::equatable-matches` implied by `-D warnings`

error: this pattern matching can be expressed using equality
  --> $DIR/all.rs:100:11
   |
LL |     while let Some(2) = o.next() {}
   |           ^^^^^^^^^^^^^^^^^^^^^^ help: try: `o.next() == Some(&2)`

error: aborting due to 17 previous errors

